core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,251 @@
+      case SpecsSnapshot(podSpecSnapshot, reservationSpecSnapshot) =>
+        if (reservationSpecSnapshot.nonEmpty) {
+          throw new NotImplementedError("ReservationSpec support not yet implemented")
+        }
+        val newPodsSpecs: Map[PodId, PodSpec] = podSpecSnapshot.map { pod => pod.id -> pod }(collection.breakOut)
+
+        val changedPodIds = podSpecs.keySet ++ newPodsSpecs.keySet
+
+        val responseFromUpdate = changedPodIds.foldLeft(ResponseBuilder.empty) { (response, podId) =>
+          SchedulerLogic.handlePodSpecUpdate(podSpecs, podRecords)(podId, newPodsSpecs.get(podId), response)
+        }
+
+        HandleSpecMessageResponse(
+          newPodsSpecs,
+          changedPodIds,
+          responseFromUpdate)
+
+      case PodSpecChange(id, newState) =>
+        val newPodSpecs = newState match {
+          case Some(podSpec) =>
+            podSpecs.updated(id, podSpec)
+          case None =>
+            podSpecs - id
+        }
+        val responseFromUpdate = SchedulerLogic.handlePodSpecUpdate(podSpecs, podRecords)(id, newState, ResponseBuilder.empty)
+
+        HandleSpecMessageResponse(
+          newPodSpecs,
+          Set(id),
+          responseFromUpdate)
+
+      case ReservationSpecChange(id, _) =>
+        throw new NotImplementedError("ReservationSpec support not yet implemented")
+    }
+  }
+
+  case class HandleMesosEventResponse(response: ResponseBuilder, changedPodIds: Set[PodId])
+  object HandleMesosEventResponse {
+    val empty = HandleMesosEventResponse(ResponseBuilder.empty, Set.empty)
+  }
+
+  def handleMesosEvent(podStatuses: Map[PodId, PodStatus], podSpecs: Map[PodId, PodSpec], cachedPendingLaunch: Set[PodId])
+                      (event: Mesos.Event): HandleMesosEventResponse = {
+    event match {
+      case offer: Mesos.Event.Offer =>
+        val changedPodIds = cachedPendingLaunch
+        val response = matchOffer(podSpecs)(offer, cachedPendingLaunch, ResponseBuilder.empty)
+        HandleMesosEventResponse(response, changedPodIds)
+
+      case Mesos.Event.StatusUpdate(taskId, taskStatus) =>
+
+        val podId = podIdFor(taskId)
+
+        if (podSpecs.contains(podId)) {
+          val newState = podStatuses.get(podId) match {
+            case Some(status) =>
+              status.copy(taskStatuses = status.taskStatuses.updated(taskId, taskStatus))
+            case None =>
+              PodStatus(podId, Map(taskId -> taskStatus))
+          }
+
+          val response = ResponseBuilder.empty.withChangeMessage(PodStatusChange(podId, Some(newState)))
+          HandleMesosEventResponse(
+            response,
+            Set(podId))
+        } else {
+          HandleMesosEventResponse.empty
+        }
+    }
+  }
+
+  /**
+    * We remove a task if it is not reachable and running, and it has no podSpec defined
+    *
+    * Should be called with the effects already applied for the specified podIds
+    * @param podIds podIds changed during the last frame
```

Comment:
you're refering to a frame here but we don't have frames anywhere. I think we should be consistent with naming...